clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2943,34 +2944,83 @@ void handleFailure(Throwable throwable) {
         final long now = time.milliseconds();
-        runnable.call(new Call("electPreferredLeaders", calcDeadlineMs(now, options.timeoutMs()),
+
+        runnable.call(new Call("electLeaders", calcDeadlineMs(now, options.timeoutMs()),
                 new ControllerNodeProvider()) {
 
             @Override
             public AbstractRequest.Builder createRequest(int timeoutMs) {
-                return new ElectPreferredLeadersRequest.Builder(
-                        ElectPreferredLeadersRequest.toRequestData(partitions, timeoutMs));
+                return new ElectLeadersRequest.Builder(ElectionType.PREFERRED, topicPartitions, timeoutMs);
             }
 
             @Override
             public void handleResponse(AbstractResponse abstractResponse) {
-                ElectPreferredLeadersResponse response = (ElectPreferredLeadersResponse) abstractResponse;
+                ElectLeadersResponse response = (ElectLeadersResponse) abstractResponse;
                 electionFuture.complete(
-                        ElectPreferredLeadersRequest.fromResponseData(response.data()));
+                        ElectLeadersResponse.fromResponseData(response.data()));
             }
 
             @Override
             void handleFailure(Throwable throwable) {
                 electionFuture.completeExceptionally(throwable);
             }
         }, now);
-        return new ElectPreferredLeadersResult(electionFuture, partitionSet);
+
+        return new ElectPreferredLeadersResult(electionFuture, topicPartitions);
     }
 
+    @Override
+    public ElectLeadersResult electLeaders(
+            final ElectionType electionType,
+            final Set<TopicPartition> topicPartitions,
+            ElectLeadersOptions options) {
+        final KafkaFutureImpl<Map<TopicPartition, Optional<Throwable>>> electionFuture = new KafkaFutureImpl<>();
+        final long now = time.milliseconds();
+        runnable.call(new Call("electLeaders", calcDeadlineMs(now, options.timeoutMs()),
+                new ControllerNodeProvider()) {
+
+            @Override
+            public AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new ElectLeadersRequest.Builder(electionType, topicPartitions, timeoutMs);
+            }
+
+            @Override
+            public void handleResponse(AbstractResponse abstractResponse) {
+                ElectLeadersResponse response = (ElectLeadersResponse) abstractResponse;
+                Map<TopicPartition, Optional<Throwable>> result = ElectLeadersResponse.electLeadersResult(response.data());
+
+                if (response.version() > 0) {
+                    Errors error = Errors.forCode(response.data().errorCode());
+                    if (error != Errors.NONE) {
+                        electionFuture.completeExceptionally(error.exception());
+                        return;
+                    }
+                } else if (topicPartitions == null && result.isEmpty()) {
```

Comment:
I'm on the fence about the need to preserve this behavior in the new API. It seems weird to allow spurious error conditions. We'll be stuck with this for a long time. I wonder if we could just document the behavior. If a user requests to elect all partitions and no response is returned, then it could mean that the user is not authorized.

Reply:
I don't have a strong opinion here. Are you thinking of just removing this `else if`?

Should the "kafka-preferred-leader-election" command handle this? Or just return the empty set there?

Reply:
Yeah, exactly. I just think the spurious auth error is more likely to cause confusion. This is even more crucial if we change the behavior of this API when partitions=null to only return the partitions that were successfully changed.